README.md
# flutterNeteaseQuickpassPlugin
flutterNeteaseQuickpassPlugin 


CHANGELOG.md
## 0.1.8

* TODO: Describe initial release.
add ios
do not use it .
do not use it .
do not use it .


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_neteasequickpass_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/integration_test/app_test.dart
// This is a basic Flutter integration test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'package:flutter_neteasequickpass_example/main.dart' as app;

void main() => run(_testMain);

void _testMain() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    app.main();

    // Trigger a frame.
    await tester.pumpAndSettle();

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_neteasequickpass/flutter_neteasequickpass.dart';
import 'dart:convert' as convert;

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _tokenMsg = '一键登录初始化中....';
  String _btnText = '一键登录初始化中..';
  bool _inited = false;


  Future<void> prefetchToken() async {
    var try_times = 0;
    do {
      try {
        Map tokenResult = await FlutterNeteasequickpass.prefetchToken();
        print("333=======");
        print(tokenResult);
        if (!mounted) return;
        if (tokenResult["code"].toString() == "ok") {
          print(" ok true ");
          setState(() {
            _tokenMsg = _tokenMsg + "\n"+ tokenResult['msg'];
            _btnText = "一键登录";
            _inited = true;
          });
          return;
        } else {
          print(" ok false ");
          setState(() {
            _tokenMsg = _tokenMsg + "\n"+ tokenResult['msg'];
            _btnText = "一键登录初始化中..重试"+try_times.toString()+"次";
          });
          if (tokenResult["code"].toString() != "net_error") {
            return;
          }
        }

      } on PlatformException catch(e){
        setState(() {
          _tokenMsg = _tokenMsg + "\n"+ e.message;
          _btnText = "一键登录初始化中..重试"+try_times.toString()+"次";
        });
        if (e.code != "net_error") {
          return;
        }
      }
      try_times += 1;
      if(try_times<3) {
        print("重试一次prefetchToken");
      }
    } while(try_times<3);
  }




  Future<void> runLogin() async {
    var try_times = 0;
    do {
      try {
        Map tokenResult = await FlutterNeteasequickpass.launchLogin();
        print(tokenResult.toString());
        if (tokenResult["code"] == "ok") {
        print("11");
          setState(() {
            _tokenMsg = _tokenMsg + "\n" + tokenResult['msg'];
          });
          print("22");
          await FlutterNeteasequickpass.quitLogin();
          print("33");
          print("ydToken is:"+tokenResult['data']['ydToken']);
          print("44");
          print("accessCode is:"+tokenResult['data']['accessCode']);
          print("55");
          return;
        } else {
          setState(() {
            _tokenMsg = _tokenMsg + "\n" + tokenResult['msg'];
          });
          if (tokenResult["code"] != "net_error" && tokenResult["code"] != "cancel") {
            return;
          }
        }
      } catch(e) {
      print(e.toString());
        //setState(() {
        //  _tokenMsg = _tokenMsg + "\n" + e.message;
        //});
        if (e.code != "net_error" && e.code != "cancel") {
          return;
        }
      }
      try_times+=1;
      if(try_times<3) {
        print("重试一次runLogin");
      }

    } while(try_times<3);
  }

  @override
  void initState() {
    print("===================================");
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String tokenMsg;
    String btnText;
    print ("222");
    try {
      var uiConfig = {
        "isHideLogo": true
      };
      String businessId =  "fa0cff55a1174fc2a63b247c8ee39da3";
      await FlutterNeteasequickpass.initialize(businessId,10,convert.jsonEncode(uiConfig));
      Map ret = await FlutterNeteasequickpass.getISPName();
      print("ISP");
      print(ret.toString());
      if (!mounted) return;
    } on PlatformException catch(e){
      print('Failed to call initialize.');
      setState(() {
        _tokenMsg = _tokenMsg+ e.message;
      });
    }
    print ("333");
    await prefetchToken();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                Text('Running on: $_tokenMsg\n'),
                FlatButton(
                    materialTapTargetSize: MaterialTapTargetSize.shrinkWrap,
                     color: Colors.white,
                    shape: RoundedRectangleBorder(
                      // side: BorderSide(
                      //   color: Colors.black12,
                      // ),
                      borderRadius: BorderRadius.circular(12),
                    ),
                    onPressed: () {
                      if (_inited) {
                          runLogin();
                      } else {

                      }

                    },
                    child: Text(_btnText),
                )
              ]
          )
        ),
      ),
    );
  }
}


lib/flutter_neteasequickpass.dart

import 'dart:async';

import 'package:flutter/services.dart';

class FlutterNeteasequickpass {
  static const MethodChannel _channel =
      const MethodChannel('flutter_neteasequickpass');

  // static Future<String> platformVersion() async {
  //   final String version = await _channel.invokeMethod('getPlatformVersion');
  //   return version;
  // }

  static Future<Map> prefetchToken() async {
    try {
      return await _channel.invokeMethod('prefetchToken');
    } on PlatformException catch (e) {
      //print("Failed to call prefetchToken: ${e.message} ");
      return {"code":e.code,"msg":e.message,"data":{"exception":e,"detail":e.details}};
    }
  }
  static Future<Map> getISPName() async {
    try {
        return await _channel.invokeMethod('getOperatorType');
    } on PlatformException catch (e) {
        //print("Failed to call prefetchToken: ${e.message} ");
        return {"code":e.code,"msg":e.message,"data":{"exception":e,"detail":e.details}};
    }
  }
  static Future<Map> initialize(String businessId,[int timeout=10,String jsonConfig="{}"]) async {
    try {
      if (businessId == "" && businessId == null) {
        return {"code":"params","msg":"businessId不能为空"};
      }
      return await _channel.invokeMethod('initialize',{"businessId":businessId,"jsonConfig":jsonConfig,"timeout":timeout});
    } on PlatformException catch (e) {
      //print("Failed to call initialize: ${e.message} ");
      return {"code":e.code,"msg":e.message,"data":{"exception":e,"detail":e.details}};
    }
  }

  static Future<Map> launchLogin() async {
    try {
      return await _channel.invokeMethod('launchLogin');
    } on PlatformException catch (e) {
      //print("Failed to call launchLogin: ${e.message} ");
      return {"code":e.code,"msg":e.message,"data":{"exception":e,"detail":e.details}};
    }
  }

  static Future<Map> quitLogin() async {
    try {
      return await _channel.invokeMethod('quitLogin');
    } on PlatformException catch (e) {
      return {"code":e.code,"msg":e.message,"data":{"exception":e,"detail":e.details}};
    }
  }

}


ios/Classes/FlutterNeteasequickpassPlugin.m
#import "FlutterNeteasequickpassPlugin.h"
#import <Flutter/Flutter.h>
#import <NTESQuickPass/NTESQuickPass.h>


@implementation FlutterNeteasequickpassPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_neteasequickpass"
            binaryMessenger:[registrar messenger]];
  FlutterNeteasequickpassPlugin* instance = [[FlutterNeteasequickpassPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];

  // if (![rootViewController isKindOfClass:[FlutterViewController class]]) {
  //   NSLog(@"expected FlutterViewController as rootViewController.");
  //   return NO;
  // }
  // FlutterViewController *flutterViewController = (FlutterViewController *)rootViewController;


}


- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    if ([@"initialize" isEqualToString:call.method]) {
        NSString *businessID = call.arguments[@"businessId"];
        NSString *jsonConfig = call.arguments[@"jsonConfig"];
        long timeout = call.arguments[@"timeout"];
        NSLog(@"businessID is %@, jsonConfig is %@, timeout is %l",businessID,jsonConfig,timeout);
        NTESQuickLoginManager * loginMgr = [NTESQuickLoginManager sharedInstance];
        BOOL ok = [loginMgr shouldQuickLogin];
        [loginMgr registerWithBusinessID:businessID];
        result(@{@"code":@"ok",@"msg":@"ok",@"data":@{}});
        // NSData *jsonData = [jsonConfig dataUsingEncoding:NSUTF8StringEncoding];
        // NSError *err;
        // NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:jsonData
        //                                                 options:NSJSONReadingMutableContainers
        //                                                   error:&err];
        // if(err) {
        //     NSLog(@"json解析失败：%@",err);
        //     return nil;
        // }

        //UIViewController *rootViewController = [[[[UIApplication sharedApplication] delegate] window] rootViewController];
        
    } else if ([@"getOperatorType" isEqualToString:call.method]) {
        result(@{@"code":@"ok",@"msg":@"ok",@"data":@{
                         @"isp":@"unknown"
        }});
    } else if ([@"prefetchToken" isEqualToString:call.method]) {
        [[NTESQuickLoginManager sharedInstance] getPhoneNumberCompletion:^(NSDictionary * _Nonnull resultDic) {
                NSNumber *boolNum = [resultDic objectForKey:@"success"];
                //WeakSelf(self);
                NSLog(@" resultDic: %@",resultDic);
                BOOL success = [boolNum boolValue];
                if (success) {
                    NSString* token = [resultDic objectForKey:@"token"];
                    weakSelf.lastToken  = token;
                    weakSelf.lastTokenTime = (long long)[[NSDate date] timeIntervalSince1970];
                    NTESQuickLoginModel *model = [[NTESQuickLoginModel alloc] init];
                    UIViewController *rootViewController = [[[[UIApplication sharedApplication] delegate] window] rootViewController];
                    model.currentVC = rootViewController;
                    model.presentDirectionType = NTESPresentDirectionPush;
                    model.authWindowPop = NTESAuthWindowPopCenter;
                    model.backgroundColor = [UIColor whiteColor];
                    model.privacyState = true;
                    model.logoImg = [UIImage imageNamed:@"app.png"];
                    [[NTESQuickLoginManager sharedInstance] setupModel:model];
                    // 设置授权登录界面model。注意：必须调用，此方法需嵌套在getPhoneNumberCompletion的回调中使用，且在CUCMAuthorizeLoginCompletion:之前调用。
                    // 电信获取脱敏手机号成功 需在此回调中拉去授权登录页面
                    result(@{@"code":@"ok",@"msg":@"ok",@"data":@{@"ydToken":token,
                        @"mobileNumber":[resultDic objectForKey:@"securityPhone"]
                    }});
                    // 移动、联通无脱敏手机号，需在此回调中拉去授权登录页面
                } else {
                     // 电信获取脱敏手机号失败
                     // 移动、联通预取号失败
                    result([FlutterError errorWithCode: @"error" message: [NSString stringWithFormat:@"获取失败:%@", [resultDic objectForKey:@"desc"]] details:@{
                        @"resultCode":[resultDic objectForKey:@"desc"]
                    }]);
                }
            }];
    } else if ([@"launchLogin" isEqualToString:call.method]) {
        if(weakSelf.lastToken == nil || ((long long)[[NSDate date] timeIntervalSince1970] - weakSelf.lastTokenTime) > 2*60*1000 ) {
            result([FlutterError errorWithCode: @"needPrefetchToken" message: @"需要调用prefetchToken" details:@{}]);
                  return;
        }
        
        [[NTESQuickLoginManager sharedInstance] CUCMCTAuthorizeLoginCompletion:^(NSDictionary * _Nonnull resultDic) {
               NSNumber *boolNum = [resultDic objectForKey:@"success"];
               BOOL success = [boolNum boolValue];
               if (success) {
                   // 取号成功，获取acessToken
                   NSLog(@" resultDic: %@",resultDic);
                   result(@{@"code":@"ok",@"msg":@"ok",@"data":@{
                        @"accessCode":[resultDic objectForKey:@"accessToken"],
                        @"ydToken":self.lastToken
                   }});
                   return;
               } else {
                    // 取号失败
                   result([FlutterError errorWithCode: @"error" message: @"获取失败" details:@{}]);
               }
             }];
    } else if ([@"quitLogin" isEqualToString:call.method]) {
        result(@{@"code":@"ok",@"msg":@"ok",@"data":@{}});
    } else {
        result([FlutterError errorWithCode: @"noSuchMethod" message: @"没有该方法" details:@{@"methodName":call.method}]);
    }
}

@end


android/src/main/java/com/loongwalk/flutter_neteasequickpass/Utils.java
package com.loongwalk.flutter_neteasequickpass;

import android.app.Activity;
import android.content.Context;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.WindowManager;
import android.widget.Toast;
import android.util.DisplayMetrics;

import java.util.Arrays;
import java.util.Map;

/**
 * Created by hzhuqi on 2019/1/29.
 */
public class Utils {

    /**
     * 将dp值转换为px值
     *
     * @param dipValue dp值
     * @return px
     */
    public static int dip2px(Context appContext, float dipValue) {
        try {
            final float scale = appContext.getResources().getDisplayMetrics().density;
            return (int) (dipValue * scale + 0.5f);
        } catch (Exception e) {
            return (int) dipValue;
        }
    }

    /**
     * 根据手机的分辨率从 px(像素) 的单位 转成为 dp
     */
    public static float px2dip(Context appContext, float pxValue) {
        final float scale = appContext.getResources().getDisplayMetrics().density;
        return pxValue / scale + 0.5f;
    }

    public static int getScreenPxWidth(Context appContext) {
        DisplayMetrics dm = appContext.getResources().getDisplayMetrics();
        return dm.widthPixels;
    }

    public static int getScreenDpWidth(Context appContext) {
        DisplayMetrics dm = appContext.getResources().getDisplayMetrics();
        return (int)(dm.widthPixels/dm.density + 0.5f);
    }

    public static int getScreenPxHeight(Context appContext) {
        DisplayMetrics dm = appContext.getResources().getDisplayMetrics();
        return dm.heightPixels;
    }

    public static int getScreenDpHeight(Context appContext) {
        DisplayMetrics dm = appContext.getResources().getDisplayMetrics();
        return (int)(dm.heightPixels/dm.density + 0.5f);
    }
}


android/src/main/java/com/loongwalk/flutter_neteasequickpass/FlutterNeteasequickpassPlugin.java
package com.loongwalk.flutter_neteasequickpass;


import java.util.HashMap;
import java.lang.System;
import androidx.annotation.NonNull;
//import android.app.Application;
import android.content.Context;
import android.util.Log;
import android.graphics.Color;
import android.util.DisplayMetrics;
import com.netease.nis.quicklogin.QuickLogin;
import com.netease.nis.quicklogin.helper.UnifyUiConfig;
import com.netease.nis.quicklogin.listener.QuickLoginPreMobileListener;
import com.netease.nis.quicklogin.listener.QuickLoginTokenListener;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterNeteasequickpassPlugin */
public class FlutterNeteasequickpassPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private QuickLogin login;
  private String lastYdToken = null;
  private long lastYdTokenTime = 0;

  private static Context appContext;
  private static boolean tokenResultReturn = false;

  public static void setTokenResultReturn(boolean b) {
      tokenResultReturn = b;
  }
  public static boolean getTokenResultReturn() {
      return tokenResultReturn;
  }

  public static HashMap<String,Object> packResult(final String code,final  String msg,final  HashMap<String,Object> data) {
    return new HashMap<String,Object>(){{
      put("code",code);
      put("msg",msg);
      put("data",data);
    }};
  }
  public static UnifyUiConfig getUiConfig() {
//    ImageView closeBtn = new ImageView(context);
//    closeBtn.setImageResource(R.drawable.close);
//    closeBtn.setScaleType(ImageView.ScaleType.FIT_XY);
//    closeBtn.setBackgroundColor(Color.TRANSPARENT);
//    RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(50, 50);
//    layoutParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, RelativeLayout.CENTER_VERTICAL);
//    layoutParams.topMargin = 30;
//    layoutParams.rightMargin = 50;
//    closeBtn.setLayoutParams(layoutParams);

//    LayoutInflater inflater = LayoutInflater.from(context);
//    RelativeLayout otherLoginRel = (RelativeLayout) inflater.inflate(R.layout.custom_other_login, null);
//    RelativeLayout.LayoutParams layoutParamsOther = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
//    layoutParamsOther.setMargins(0, 0, 0, Utils.dip2px(context, 130));
//    layoutParamsOther.addRule(RelativeLayout.CENTER_HORIZONTAL);
//    layoutParamsOther.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
    //otherLoginRel.setLayoutParams(layoutParamsOther);
//    ImageView wx = otherLoginRel.findViewById(R.id.weixin);
//    ImageView qq = otherLoginRel.findViewById(R.id.qq);
//    ImageView wb = otherLoginRel.findViewById(R.id.weibo);
//    wx.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "微信登录", Toast.LENGTH_SHORT).show();
//      }
//    });
//    qq.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "qq登录", Toast.LENGTH_SHORT).show();
//
//      }
//    });
//    wb.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "微博登录", Toast.LENGTH_SHORT).show();
//      }
//    });
    int X_OFFSET = 0;
    int BOTTOM_OFFSET = 0;
    UnifyUiConfig uiConfig = new UnifyUiConfig.Builder()
            // 状态栏
//            .setStatusBarColor(Color.RED)
//            .setStatusBarDarkColor(true)
            // 设置导航栏
            .setHideNavigation(true)
//            .setNavigationTitle("一键登录/注册")
//            //.setNavigationTitleColor(Color.RED)
//            .setNavigationBackgroundColor(Color.BLUE)
//            .setNavigationIcon("yd_checkbox_checked2")
//            .setNavigationBackIconWidth(25)
//            .setNavigationBackIconHeight(25)
            //.setHideNavigation(false)
            // 设置logo
            //.setLogoIconName("ico_logo")
//            .setLogoWidth(70)
//            .setLogoHeight(70)
//            .setLogoXOffset(X_OFFSET)
//            .setLogoTopYOffset(50)
//            .setHideLogo(false)
            //手机掩码
            .setMaskNumberColor(Color.RED)
            .setMaskNumberSize(15)
            .setMaskNumberXOffset(X_OFFSET)
            .setMaskNumberTopYOffset(120)
            .setMaskNumberBottomYOffset(BOTTOM_OFFSET)
            // 认证品牌
            .setSloganSize(15)
            .setSloganColor(Color.RED)
            .setSloganXOffset(X_OFFSET)
            .setSloganTopYOffset(200)
            .setSloganBottomYOffset(BOTTOM_OFFSET)
            // 登录按钮
            .setLoginBtnText("一键登录")
            .setLoginBtnTextColor(Color.WHITE)
            //.setLoginBtnBackgroundRes("btn_shape_login_onepass")
            .setLoginBtnWidth(100)
            .setLoginBtnHeight(45)
            .setLoginBtnTextSize(15)
            //.setLoginBtnXOffset(X_OFFSET)
            //.setLoginBtnTopYOffset(250)
            .setLoginBtnBottomYOffset(BOTTOM_OFFSET)
            // 隐私栏
            .setPrivacyTextStart("登录即同意")
            .setProtocolText("服务条款一")
            .setProtocolLink("https://www.baidu.com")
            .setProtocol2Text("服务条款二")
            .setProtocol2Link("https://www.baidu.com")
            //.setPrivacyTextEnd("且授权易盾一键登录SDK使用本机号码")
            .setPrivacyTextColor(Color.RED)
            .setPrivacyProtocolColor(Color.GREEN)
//                .setHidePrivacyCheckBox(false)
            .setPrivacyXOffset(X_OFFSET)
            .setPrivacyState(true)
            .setPrivacySize(12)
//                .setPrivacyTopYOffset(510)
            .setPrivacyBottomYOffset(20)
            .setPrivacyTextGravityCenter(true)
            .setCheckedImageName("yd_checkbox_checked2")
            .setUnCheckedImageName("yd_checkbox_unchecked2")
            // 协议详情页导航栏
            .setProtocolPageNavTitle("一键登录服务条款")
            //.setProtocolPageNavBackIcon("yd_checkbox_checked")
            .setProtocolPageNavColor(Color.BLUE)

//                .setBackgroundImage("bg1")
            // 自定义控件
//            .addCustomView(otherLoginRel, "relative", UnifyUiConfig.POSITION_IN_BODY, null)
//            .addCustomView(closeBtn, "close_btn", UnifyUiConfig.POSITION_IN_TITLE_BAR, new LoginUiHelper.CustomViewListener() {
//              @Override
//              public void onClick(Context context, View view) {
//                Toast.makeText(context, "点击了右上角X按钮", Toast.LENGTH_SHORT).show();
//              }
//            })
            .build(appContext);
    return uiConfig;
  }


  public static UnifyUiConfig getDialogUiConfig() {
//    ImageView closeBtn = new ImageView(context);
//    closeBtn.setImageResource(R.drawable.close);
//    closeBtn.setScaleType(ImageView.ScaleType.FIT_XY);
//    closeBtn.setBackgroundColor(Color.TRANSPARENT);
//    RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(60, 60);
//    layoutParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, RelativeLayout.CENTER_VERTICAL);
//    layoutParams.topMargin = 30;
//    layoutParams.rightMargin = 50;
//    closeBtn.setLayoutParams(layoutParams);

//    LayoutInflater inflater = LayoutInflater.from(context);
//    RelativeLayout otherLoginRel = (RelativeLayout) inflater.inflate(R.layout.custom_other_login, null);
//    RelativeLayout.LayoutParams layoutParamsOther = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
//    layoutParamsOther.setMargins(0, 0, 0, Utils.dip2px(context, 100));
//    layoutParamsOther.addRule(RelativeLayout.CENTER_HORIZONTAL);
//    layoutParamsOther.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
//    otherLoginRel.setLayoutParams(layoutParamsOther);
//    ImageView wx = otherLoginRel.findViewById(R.id.weixin);
//    ImageView qq = otherLoginRel.findViewById(R.id.qq);
//    ImageView wb = otherLoginRel.findViewById(R.id.weibo);
//    wx.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "微信登录", Toast.LENGTH_SHORT).show();
//      }
//    });
//    qq.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "qq登录", Toast.LENGTH_SHORT).show();
//
//      }
//    });
//    wb.setOnClickListener(new View.OnClickListener() {
//      @Override
//      public void onClick(View v) {
//        Toast.makeText(context, "微博登录", Toast.LENGTH_SHORT).show();
//      }
//    });
    int X_OFFSET = 0;
    DisplayMetrics dm = new DisplayMetrics();
    int screenWidth = Utils.getScreenDpWidth(appContext);
    int screenHeight = Utils.getScreenDpHeight(appContext);
    int dialogWidth = (int) ( screenWidth * 0.85);
    int dialogHeight = (int) ( screenHeight * 0.5);
    UnifyUiConfig uiConfig = new UnifyUiConfig.Builder()
            // 设置导航栏
//            .setNavigationTitle("一键登录/注册")
//            .setNavigationTitleColor(Color.RED)
//            .setNavigationBackgroundColor(Color.BLUE)
//            .setNavigationIcon("yd_checkbox_checked")
            .setHideNavigation(true)
            // 设置logo
//            .setLogoIconName("ico_logo")
//            .setLogoWidth(50)
//            .setLogoHeight(50)
//            .setLogoXOffset(X_OFFSET)
//            .setLogoTopYOffset(15)
////                .setLogoBottomYOffset(300)
            .setHideLogo(true)
            //手机掩码
            .setMaskNumberColor(Color.BLACK)
            .setMaskNumberSize(15)
            .setMaskNumberTopYOffset(100)
            .setMaskNumberXOffset(X_OFFSET)
            // 认证品牌
//            .setSloganSize(15)
//            .setSloganColor(Color.BLACK)
//            .setSloganTopYOffset(70)
//            .setSloganXOffset(X_OFFSET)
            // 登录按钮
            .setLoginBtnText("一键登录")
            .setLoginBtnTextColor(Color.BLACK)
            .setLoginBtnBackgroundRes("btn_shape_login_onepass")
            .setLoginBtnWidth(100)
            .setLoginBtnHeight(40)
            .setLoginBtnTextSize(15)
            .setLoginBtnXOffset(X_OFFSET)
            .setLoginBtnTopYOffset(150)
            // 隐私栏
            //.setPrivacyTextStart("登录即同意")
            //.setProtocolText("服务条款一")
            //.setProtocolLink("https://www.baidu.com")
            //.setProtocol2Text("服务条款二")
            //.setProtocol2Link("https://www.baidu.com")
            .setProtocolPageNavTitle("")
            .setProtocolPageNavHeight(0)
            .setProtocolPageNavTitleSize(0)
            .setPrivacyTextEnd("且授权Bee.com使用本机号码")
            .setPrivacyMarginRight((int)(screenWidth*0.1))
            //.setPrivacyMarginLeft(30)
            .setPrivacyTextColor(Color.GRAY)
            .setPrivacyProtocolColor(Color.BLUE)
            .setHidePrivacyCheckBox(true)
            .setPrivacyXOffset((int)(screenWidth*0.1))
            .setPrivacyState(true)
            .setPrivacySize(12)
            .setPrivacyBottomYOffset((int)(screenHeight*0.05))
            .setPrivacyTextGravityCenter(true)
            //.setCheckedImageName("yd_checkbox_checked")
            //.setUnCheckedImageName("yd_checkbox_unchecked")
            // 协议详情页导航栏
            //.setProtocolPageNavTitle("易盾一键登录SDK服务条款")
            //.setProtocolPageNavBackIcon("yd_checkbox_checked")
            //.setProtocolPageNavColor(Color.BLUE)

//            .addCustomView(otherLoginRel, "relative", UnifyUiConfig.POSITION_IN_BODY, null)
//            .addCustomView(closeBtn, "close_btn", UnifyUiConfig.POSITION_IN_TITLE_BAR, new LoginUiHelper.CustomViewListener() {
//              @Override
//              public void onClick(Context context, View view) {
//                Toast.makeText(context.getApplicationContext(), "点击了右上角X按钮", Toast.LENGTH_SHORT).show();
//              }
//            })
            .setDialogMode(true, dialogWidth, dialogHeight, 0, 0, false)
            .build(appContext);
    return uiConfig;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_neteasequickpass");
    channel.setMethodCallHandler(this);
    appContext = flutterPluginBinding.getApplicationContext();

  }




  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result2) {
    final Result result = result2;
    final FlutterNeteasequickpassPlugin that = this;
    Log.i("flutter","call name is "+call.method);
    if (call.method.equals("prefetchToken")) {

      if(login==null) {
          result.error("uninitialize","未初始化",null);
          return;
      }
      //Log.i("flutter","call prefetchToken");
      login.prefetchMobileNumber(new QuickLoginPreMobileListener() {
        @Override
        public void onGetMobileNumberSuccess(final String YDToken, final String mobileNumber) {
          //Log.i("flutter", String.format("yd token is:%s mobileNumber is:%s", YDToken, mobileNumber));
          lastYdToken = YDToken;
          lastYdTokenTime = System.currentTimeMillis();

          result.success(packResult("ok","ok",new HashMap<String,Object>(){{
            put("ydToken",YDToken);
            put("mobileNumber",mobileNumber);
          }}));
        }
        @Override
        public void onGetMobileNumberError(final String YDToken, final String msg) {
            //Log.i("flutter","  call prefetchToken failed !!"+msg);
            if(msg.contains("网络") || msg.contains("java.net") ) {
                result.error("net_error",msg,new HashMap<String,Object>(){{
                  put("ydToken",YDToken);
                }});
            } else {
                result.error("error",msg,new HashMap<String,Object>(){{
                  put("ydToken",YDToken);
                }});
            }
        }
      });
    } else if (call.method.equals("getOperatorType")) {
      if(login==null) {
        result.error("uninitialize","未初始化",null);
        return;
      }
//      1:      电信
//              * 2:      移动
//              * 3:      联通
//              * 5:      未知
      String ispName = "unknown";
      int ret = login.getOperatorType(appContext);
      switch(ret) {
        case 1:
          ispName = "ChinaTelecom";
          break;
        case 2:
          ispName = "ChinaMobile";
          break;
        case 3:
          ispName = "ChinaUnicom";
          break;
      }
      final String isp = ispName;
      result.success(packResult("ok","ok",new HashMap<String,Object>(){{
        put("isp",isp);
      }}));
      return;
    } else if (call.method.equals("launchLogin")) {
        if(login==null) {
          result.error("uninitialize","未初始化",null);
          return;
        }
        if(lastYdToken == null || (System.currentTimeMillis() - lastYdTokenTime) > 2*60*1000 ) {
          result.error("needPrefetchToken","需要调用prefetchToken",null);
          return;
        }

        that.setTokenResultReturn(false);
        login.onePass(new QuickLoginTokenListener() {
          @Override
          public void onGetTokenSuccess(final String YDToken, final String accessCode) {
            Log.i("flutter", String.format("yd token is:%s accessCode is:%s", YDToken, accessCode));
            //tokenValidate(YDToken, accessCode, true);
            if(that.getTokenResultReturn()) {
              return;
            }
            that.setTokenResultReturn(true);
            result.success(packResult("ok","ok",new HashMap<String,Object>(){{
              put("ydToken",YDToken);
              put("accessCode",accessCode);
            }}));
          }

          @Override
          public void onGetTokenError(final String YDToken, String msg) {
            Log.i("flutter", "获取运营商授权码失败:" + msg);
            if(that.getTokenResultReturn()) {
              return;
            }
            that.setTokenResultReturn(true);
            result.error("error",msg,new HashMap<String,Object>(){{
              put("ydToken",YDToken);
            }});
          }
          @Override
          public void onCancelGetToken() {
            Log.i("flutter", "用户取消登录");
            if(that.getTokenResultReturn()) {
              return;
            }
            that.setTokenResultReturn(true);
            result.error("cancel","用户取消登录",new HashMap<String,Object>(){{
            }});
          }
        });
        Log.i("flutter","call launchLogin");
    } else if (call.method.equals("quitLogin")) {
        if(login==null) {
          result.error("uninitialize","未初始化",null);
          return;
        }
        Log.i("flutter","call qiltLogin");
        login.quitActivity();
        result.success(packResult("ok","ok",new HashMap<String,Object>(){{
        }}));
    } else if (call.method.equals("initialize")) {
      final String businessId = call.argument("businessId");
      final String jsonConfigStr = call.argument("jsonConfig");
      final int timeout = call.argument("timeout");
      if(login!=null) {
        result.success(packResult("ok","ok",new HashMap<String,Object>(){{
          put("businessId",businessId);
        }}));
        return;
      }

      Log.i("flutter","call initialize businessId is "+businessId+ "  jsonConfigStr "+jsonConfigStr+ " timeout: "+timeout);
      login = QuickLogin.getInstance(appContext, businessId);// BUSINESS_ID为从易盾官网申请的业务id
      login.setPrefetchNumberTimeout(timeout); // 设置预取号超时时间，单位s
      login.setFetchNumberTimeout(timeout);
      login.setUnifyUiConfig((new JsonConfig(appContext)).buildUiConfig(jsonConfigStr));
      result.success(packResult("ok","ok",new HashMap<String,Object>(){{
          put("businessId",businessId);
      }}));
    } else {
      final String methodName = call.method;
      result.error("nomethod","没有该方法",new HashMap<String,Object>(){{
        put("methodName",methodName);
      }});

    }

  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
